### ISSUE TITLE
Change Request: Expose the active language ID to rules

### ISSUE THREAD

**Issue by sindresorhus** [eslint/eslint#21343]
### ESLint version

v10.9.1

### What problem do you want to solve?

Rules can declare `meta.languages` but the active language ID is not available through the rule context.

For example, `json/json`, `json/jsonc`, and `json/json5` normally have identical `context.languageOptions` and use the same `SourceCode` implementation. A rule therfore cannot reliably determine which JSON language is active.

The filename extension cannot identify the language reliably. Flat config can assign any language to any filename pattern, files can be linted from stdin using a virtual filename, processors can create virtual files, and `.json` files may intentionally use JSONC or JSON5. The configured language is the source of truth, not the filename.

This matters when behavior differs by language. An autofix may emit `\v` or `\0` for JSON5, but those escapes are invalid in JSON and JSONC. Today, a rule must guess from the filename, require a redundant option, or omit otherwise valid behaviour.

### What do you think is the correct solution?

Expose the active runtime language identifier as a read-only rule context property:

```js
context.languageId === 'json/json5'
```

This should use the same normalized identifier semantics as `meta.languages`, ideally using the plugin's `meta.namespace` so configuring a plugin under an alias does not change rule behavior.

A string identifier is preferable to exposing the resolved language object. This information should not be added to `languageOptions` because it describes ESLint's execution context rather than parser or language configuration.

### Participation

- [ ] I am willing to submit a pull request for this change.

### AI acknowledgment

- [x] I did not use AI to generate this issue report.
- [ ] (If the above is not checked) I have reviewed the AI-generated content before submitting.

**Comment by bytedoe** [comment]
I’m willing to open a PR for this.

**Comment by lumirlumir** [comment]
I’m linking to a similar issue I raised in the past about exposing `context.language`, similar to the proposal here... However, it was closed as not planned at the time, with the following response:

> Thanks for the suggestion. We purposely don't expose `language` inside of rules because we don't want rules altering how they behave based on the language used to parse it. Any number of languages can produce the same AST and so exposing the name of the language potentially makes the rules more likely to break.
>
> Rules are designed to function based only on the AST produced and the `languageOptions` specified. Anything more than that and the behavior gets too unpredictable and can be frustrating for users.

https://github.com/eslint/rewrite/issues/221

**Comment by DMartens** [comment]
In case this problem only affects JSON, we could add a `mode` to the JSON `Document`.
This would be akin to `sourceType` on a JavaScript `Program` and would not expose language internals as lumir mentioned.
